### PR TITLE
Fix type check for `ordered` types

### LINKED ISSUE #29643: Fix type checking for `ordered` types
**Description:**
Refer https://github.com/ballerina-platform/ballerina-spec/issues/787

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by lasinicl**
## Purpose
$title

Fixes #29643

## Approach
Unions of different types are not ordered types. ex:int|float

## Check List 
- [X] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [X] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 1]
```diff
@@ -5769,8 +5769,32 @@ private void checkArrayLibSortFuncArgs(BLangInvocation iExpr) {
...
+                return;
+            }
+        } else {
+            BLangLambdaFunction keyLambdaFunction = (BLangLambdaFunction) keyFunction;
+            pos = keyLambdaFunction.function.pos;
+            returnType = keyLambdaFunction.function.type.getReturnType();
+        }
+
+        if (!types.isOrderedType(returnType, false)) {
+            dlog.error(pos, DiagnosticErrorCode.INVALID_SORT_FUNC_RETURN_TYPE, returnType);
```
Comment (lasinicl) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:5797]: Added back the code which was originally there

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 2]
```diff
@@ -2657,6 +2660,20 @@ private boolean checkUnionHasSameFiniteType(LinkedHashSet<BType> memberTypes, BT
         return false;
     }
 
+    private boolean checkValueSpaceHasSameType(BFiniteType finiteType, BType baseType) {
+        if (baseType.tag == TypeTags.FINITE) {
+            return finiteType == baseType;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:2665]: Won't work for something like the following, right?

```ballerina
type Foo 1|2;
type Bar 10|11;

public function main() {
    Foo f = 1;
    Bar b = 10;

    boolean x = f < b;
}
```

Comment (lasinicl) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: Fixed in 4bbf165

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 3]
```diff
@@ -2657,6 +2660,20 @@ private boolean checkUnionHasSameFiniteType(LinkedHashSet<BType> memberTypes, BT
         return false;
     }
 
+    private boolean checkValueSpaceHasSameType(BFiniteType finiteType, BType baseType) {
+        if (baseType.tag == TypeTags.FINITE) {
+            return true;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:2665]: Please check the following. It should fail at compile-time. For whatever that was not supported, let's go with the most restrictive approach for alpha4 and relax the restriction in a separate PR.

We should make sure whatever correct comparison that worked previously should work though.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 4]
```diff
@@ -2624,7 +2624,10 @@ public Boolean visit(BTypedescType t, BType s) {
 
         @Override
         public Boolean visit(BFiniteType t, BType s) {
-
+            if (inOrderedType) {
```
[the remark this answers is not in the document]
Comment (lasinicl) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:2627]: Created an issue https://github.com/ballerina-platform/ballerina-lang/issues/29657

**Review by MaryamZi — APPROVED** [review]
@lasinicl, as discussed offline let's clarify what the runtime behaviour should be when the result is UN.

**Pull request merged**